dali/imgcodec/image_decoder_test.cc

```diff
@@ -49,49 +65,202 @@ struct ImageBuffer {
   ImageSource src;
   explicit ImageBuffer(const std::string &filename)
   : buffer(read_file(filename))
-  , src(ImageSource::FromHostMem(buffer.data(), buffer.size())) {}
+  , src(ImageSource::FromHostMem(buffer.data(), buffer.size(), filename)) {}
 };
 
 struct test_sample {
-  test_sample(std::string img_path, std::string npy_ref_path)
+  test_sample(std::string img_path, std::string npy_path, std::string npy_ycbcr_path,
+              std::string npy_gray_path)
       : image(img_path),
-        ref(numpy::ReadTensor(FileStream::Open(npy_ref_path, false, false).get())) {}
+        ref(numpy::ReadTensor(FileStream::Open(npy_path, false, false).get())),
+        ref_ycbcr(numpy::ReadTensor(FileStream::Open(npy_ycbcr_path, false, false).get())),
+        ref_gray(numpy::ReadTensor(FileStream::Open(npy_gray_path, false, false).get())) {}
+
+  const Tensor<CPUBackend> &GetRef(DALIImageType format) {
+    if (format == DALI_YCbCr) {
+      return ref_ycbcr;
+    } else if (format == DALI_GRAY) {
+      return ref_gray;
+    } else {
+      assert(format == DALI_RGB || format == DALI_ANY_DATA);
+      return ref;
+    }
+  }
 
   ImageBuffer image;
   Tensor<CPUBackend> ref;
+  Tensor<CPUBackend> ref_ycbcr;
+  Tensor<CPUBackend> ref_gray;
 };
 
+cv::Mat rgb2bgr(const cv::Mat &img) {
+  cv::Mat bgr;
+  cv::cvtColor(img, bgr, cv::COLOR_RGB2BGR);
+  return bgr;
+}
+
+template <typename T>
+cv::Mat GetCvMat(TensorView<StorageCPU, const T> v, DALIImageType color_fmt) {
+  static_assert(std::is_same<T, uint8_t>::value, "Only uint8 images supported for now");
+  cv::Mat v_mat(v.shape[0], v.shape[1], color_fmt == DALI_GRAY ? CV_8UC1 : CV_8UC3,
+                (void *)v.data);  // NOLINT
+  return rgb2bgr(v_mat);
+}
+
+cv::Mat GetRefCvMat(test_sample& sample, DALIImageType color_fmt) {
+  void *refdata = nullptr;
+  TensorShape<> ref_sh;
+  if (color_fmt == DALI_YCbCr) {
+    refdata = sample.ref_ycbcr.raw_mutable_data();
+    ref_sh = sample.ref_ycbcr.shape();
+  } else if (color_fmt == DALI_GRAY) {
+    refdata = sample.ref_gray.raw_mutable_data();
+    ref_sh = sample.ref_gray.shape();
+  } else {
+    refdata = sample.ref.raw_mutable_data();
+    ref_sh = sample.ref.shape();
+  }
+  cv::Mat vref_mat(ref_sh[0], ref_sh[1], color_fmt == DALI_GRAY ? CV_8UC1 : CV_8UC3, refdata);
+  return rgb2bgr(vref_mat);
+}
+
+template <typename T>
+void DumpImages(TensorView<StorageCPU, const T> v,
+                test_sample& sample,
+                DALIImageType color_fmt) {
+  // not implemented
+}
+
+void DumpImages(TensorView<StorageCPU, const uint8_t> v,
+                test_sample& sample,
+                DALIImageType color_fmt) {
+  auto out = GetCvMat<uint8_t>(v, color_fmt);
+  auto ref = GetRefCvMat(sample, color_fmt);
+  cv::Mat diff;
+  cv::absdiff(out, ref, diff);
+  cv::imwrite("/tmp/img_out.bmp", out);
+  cv::imwrite("/tmp/img_ref.bmp", ref);
+  cv::imwrite("/tmp/img_diff.bmp", diff);
```

Comment:
I'm not a fan of this - for once, it will only store the last error. Also, it dumps some stuff into `/tmp` directory which is quite unexpected - I think the other tests just write to current dir (if indeed they write anything).
```suggestion
  static int counter = 0;
  int n = counter++;
  char out_name[64], ref_name[64], diff_name[64];
  snprintf(out_name, 64, "img_%04d_out.bmp", n);
  snprintf(ref_name, 64, "img_%04d_ref.bmp", n);
  snprintf(diff_name, 64, "img_%04d_diff.bmp", n);
  cv::imwrite(out_name, out);
  cv::imwrite(ref_name, ref);
  cv::imwrite(diff_name, diff);
```